### PR TITLE
`LazyIntRope` specialization for `string_to_inum` primitive

### PR DESCRIPTION
**Pull request by nirvdrum**
This PR takes care of two related issues. Most of the work went into a clean-up pass on `ConvertBytes`. That process simplified Ruby exceptions thrown from within `ConvertBytes`, which used to require having a reference to both a Ruby string and its underlying rope. With that clean-up in place, the `string_to_inum` primitive was updated to work with ropes directly, which makes specializing on rope metadata easier.

The PR also adds a new specialization from the `string_to_inum` primitive around `LazyIntRope`. If the base is either `0` (signifying base auto-discovery; `Kernel#Integer` default) or `10` (`String#to_i` default), then we can simply read the integer value out of the `LazyIntRope` without having to look at its string bytes at all. While that process admittedly sounds a bit contrived, it is a real case that [occurs in Rack](https://github.com/rack/rack/blob/1741c580d71cfca8e541e96cc372305c8892ee74/lib/rack/request.rb#L233-L254). Here, the port may be specified as an integer, which is turned into a String  to function as a server header, which is then converted to an integer (with `Kernel#integer`) -- presumably for data validation, and then turned back into a string via the `"#{host}:#{port}"` string interpolation.

I believe there are opportunities for other specializations that may be able to return results efficiently. E.g., as far as I can tell, any string with a `CR_VALID` coderange cannot be converted into an integer, and so would either turn into `0` or an exception, without ever having to actually look at its contents. I'm leaving that for future work because I think if we add a `ConvertBytesNodes` class, we can optimize common cases nicely.

Currently, all of the machinery for converting a string to an integer is behind a boundary. That code needs to handle base conversions `2..36` and has a logic switch if the base is `0` or negative. Looking through a real world application at Shopify, only two bases were ever used: `0` (the `Kernel#Integer` default) and `10` (the `String#to_i` default). It may makes sense to also optimize bases 2, 8, and 16 since there are special prefixes for those, although with some supporting method (`String#hex` and `String#oct`). `ConvertBytes` is also used for methods other than `string_to_inum`, but I haven't investigated those beyond working out exceptional paths.

### CODE REVIEW

**Comment by chrisseaton** [comment]
Great work, Kevin! A rare example of a non-local optimisation in Truffle. We should Tweet about this one as it's a very powerful but simple optimisation with a real-world motivation.

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 1]
```diff
@@ -5105,16 +5106,42 @@ protected boolean indexAtEitherBounds(RubyString string, int index) {
...
+
+        @Specialization(guards = {
+                "isLazyIntRope(rope)",
+                "doesNotRequireBaseConversion(fixBase)"
+        })
+        protected Object stringToInumIntRope(Rope rope, int fixBase, boolean strict, boolean raiseOnError) {
+            return ((LazyIntRope) rope).getValue();
+        }
 
         @Specialization
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java:5133]: We need a negated guard here, so this specialization is never taken fro a LazyIntRope with base=0||10

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 2]
```diff
@@ -5105,16 +5106,42 @@ protected boolean indexAtEitherBounds(RubyString string, int index) {
...
+                @Cached BranchProfile exceptionProfile) {
+            if (bytesNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                bytesNode = insert(RopeNodes.BytesNode.create());
+            }
+
+            if (fixnumOrBignumNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                fixnumOrBignumNode = insert(new FixnumOrBignumNode());
+            }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java:5144]: `@Cached` is preferred to the lazy pattern if only used in one specialization, could you move those 2 nodes back to `@Cached`?

Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/StringNodes.java:5144]: Will do. Sorry about that. I had three specializations at one point, but backed one out for now.

**Review by eregon — APPROVED** [review]
Nice! So `integer.to_s.to_i` is basically no-op now on TruffleRuby :)

**Comment by chrisseaton** [comment]
Great - now let's exploit with Tweets etc.

**Pull request merged**